trackpy/linking/utils.py

```diff
@@ -38,23 +38,27 @@ def coords_from_df(df, pos_columns, t_column):
     t_column to be of integer type. Float-typed integers are also accepted.
 
     Empty frames will be returned as empty arrays of shape (0, ndim)."""
+
+    # This implementation is much faster than using DataFrame.groupby.
+
     ndim = len(pos_columns)
-    grouped = iter(df.groupby(t_column))  # groupby sorts by default
-
-    # get the first frame to learn first frame number
-    cur_frame, frame = next(grouped)
-    cur_frame = int(cur_frame)
-    yield cur_frame, frame[pos_columns].values
-    cur_frame += 1
-
-    for frame_no, frame in grouped:
-        frame_no = int(frame_no)
-        while cur_frame < frame_no:
-            yield cur_frame, np.empty((0, ndim))
-            cur_frame += 1
-
-        yield cur_frame, frame[pos_columns].values
-        cur_frame += 1
+    times = df[t_column].values
+    pos = df[pos_columns].values
+
+    idxs = np.argsort(times, kind="mergesort")  # i.e. stable
+    times = times[idxs]
+    pos = pos[idxs]
+
+    unique_times, time_counts = np.unique(times, return_counts=True)
+    pos_by_frame = np.split(pos, np.cumsum(time_counts)[:-1])
+
+    idx = 0
+    for time in range(unique_times[0], unique_times[-1] + 1):
+        if time == unique_times[idx]:
+            yield time, pos_by_frame[idx]
+            idx += 1
+        else:
+            yield time, np.empty((0, ndim))
```

Comment:
Nice! From my read through it looks like this is really equivalent in terms of results and data types, which is of course borne out by @caspervdw 's tests. Pandas helped trackpy avoid premature optimization, but it's good to see it replaced in one more algorithm.